### PR TITLE
[Core] Support offloading KV cache to CPU

### PR DESCRIPTION
**Pull request by ApostaC**
## An implementation for CPU KV cache offloading (#7697)

TL; DR: **CPU offloading is better than prefix caching in our benchmark**, we also found that the evictor can be optimized to save 10-30% of the runtime.

This PR is for fixing the DCO issue for the Kuntai's original CPU offloading PR #9682 . It also contains new CUDA kernels to improve the KV cache offloading performance.

## End-to-end benchmarking results:

A long document QA workload (see `benchmarks/benchmark_long_document_qa.py`) running on A100-40G-SXM GPU. The GPU can cache 8 documents and the CPU can cache 30 documents. 

image

(Following are the original data for the above figure)
| Num documents 	| vLLM   	| vLLM w/ prefix caching 	| vLLM w/ prefix caching + CPU offloading 	|
|---------------	|--------	|------------------------	|-----------------------------------------	|
| 8             	| 13.66  	| 0.49                   	| 0.5                                     	|
| 16            	| 27.28  	| 7.22                   	| 2.3                                     	|
| 32            	| 54.54  	| 49.96                  	| 17.26                                   	|
| 64            	| 109.27 	| 126.08                 	| 110.96                                  	|

## New kernel implementation microbenchmark

The numbers are collected on A100-40GB-SXM GPUs

| # of pages          	| New implementation 	| Old implementation 	|
|---------------------	|--------------------	|--------------------	|
| Swap in 1 page      	| 1.21 GB / second   	| 1.21 GB / second   	|
| Swap in 1250 pages  	| 12.6 GB / second   	| 3.11 GB / second   	|
| Swap out 1 page     	| 1.22 GB / second   	| 1.21 GB / second   	|
| Swap out 1250 pages 	| 12.5 GB / second   	| 3.12 GB / seocond  	|

The new kernel can achieve 4x better throughput than the old `swap_block` implementation.
Also, it won't decrease the performance when the number of pages are small.

### Potential improvement:
Currently, the `swap_block` is invoked once per layer. If we can aggregate the copy of all the layers into one kernel, the throughput of copying 1 page will also achieve >10GB/s.

## Implementation

This PR has much less features compared to #8694, but it is really minimum and creates very little core change. So I guess we can use this PR to enable CPU KV cache offloading first, and then focus on disk.

The key idea of this implementation is to maintain those allocated blocks that didn't hit the cache, and constantly copy them into CPU after each scheduler step.

Here is the flow diagram
image

This idea is borrowed from ConServe (paper link: https://arxiv.org/abs/2410.01228), based on the assumption that the CPU-GPU bandwidth is much higher than GPU KV cache generation throughput. Thanks Yifan for this idea.

### CODE REVIEW

File: tests/kernels/test_cache.py [hunk 1]
```diff
@@ -362,7 +362,7 @@ def test_swap_blocks(
     block_mapping = list(zip(src_blocks, dst_blocks))
     block_mapping_tensor = torch.tensor(block_mapping,
                                         dtype=torch.int64,
-                                        device="cpu").view(-1, 2)
+                                        device=device).view(-1, 2)
```
Comment (KuntaiDu) [inline tests/kernels/test_cache.py:365]: Is this because that this tensor need to be accessed by the new CUDA memcpy kernel?

Comment (ApostaC) [reply tests/kernels/test_cache.py:365]: Yes. The new `paged_copy` kernel need to access the block mapping from GPU.

File: vllm/core/block_manager.py [hunk 2]
```diff
@@ -508,3 +523,19 @@ def get_num_cached_tokens(self, seq: Sequence) -> int:
         cached in the block manager for the sequence.
         """
         return self._computed_blocks_tracker.get_num_cached_tokens(seq)
+
+    def get_and_reset_swaps(self,
```
Comment (lixiaobai09) [inline vllm/core/block_manager.py:527]: This function seems like not get the real physical block ID from `get_physical_block_id`? Especially for CPU PrefixCachingBlockAllocator, whose start ID is not zero.

Comment (ApostaC) [reply vllm/core/block_manager.py:527]: If I understand correctly, this function should not return the physical block IDs because the `get_physical_block_id` will be called in `block_manager.swap_in()`/`block_manager.swap_out()` later. 

Call chain is: `scheduler._swap_in() --> block_manager.swap_in() --> block_allocator.get_physical_block_id()` (similar for swapping out).

(Let me know if my understanding is incorrect and I will fix it asap, thanks!)

File: vllm/core/scheduler.py [hunk 3]
```diff
@@ -1200,6 +1214,21 @@ def _schedule_chunked_prefill(self) -> SchedulerOutputs:
...
+
+        blocks_to_copy = running_scheduled.blocks_to_copy
+        blocks_to_copy.extend(swapped_in.blocks_to_copy)
+
+        blocks_to_swap_in = swapped_in.blocks_to_swap_in
+        blocks_to_swap_out = running_scheduled.blocks_to_swap_out
+
+        # NOTE(Kuntai): extend the swapping list for CPU offloading
+        new_swap_out, new_swap_in = \
+                self.block_manager.get_and_reset_swaps(time.time())
```
Comment (lixiaobai09) [inline vllm/core/scheduler.py:1226]: However, the `get_and_reset_swaps` is called directly here without `get_physical_block_id`. I think these block IDs are sent to the cache engine directly later, so not the real physical block IDs.

Comment (ApostaC) [reply vllm/core/scheduler.py:1226]: Got it! I double-checked the logic and you are right. Just pushed another commit to fix the issue and update the docstring. Thanks for the catch!

File: vllm/core/block/cpu_offloading_block_allocator.py [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+                               PrefixCachingBlockAllocator] = {  # type: ignore
+                                   Device.CPU: cpu_block_allocator,
+                                   Device.GPU: gpu_block_allocator
+                               }
+        """
+        GPU block should only be in one of the following three status:
+          uncached: allocated blocks that didn't hit any cache
+          cached: allocated blocks that are cached, either in GPU or in CPU
+          free: the blocks are not allocated by block allocator
+        This implementation aims to transform uncacherd blocks to cached blocks
```
Comment (DearPlanet) [inline vllm/core/block/cpu_offloading_block_allocator.py:114]: uncacherd blocks -> uncached blocks?

Comment (ApostaC) [reply vllm/core/block/cpu_offloading_block_allocator.py]: Thanks for the catch! Fixed!

File: vllm/engine/arg_utils.py [hunk 5]
```diff
@@ -413,6 +414,17 @@ def add_cli_args(parser: FlexibleArgumentParser) -> FlexibleArgumentParser:
...
+            '--block-allocator',
+            type=str,
+            default='CpuGpuBlockAllocator',
+            choices=['CpuGpuBlockAllocator', 'CpuOffloadingBlockAllocator'],
+            help='The block allocator that vLLM uses. Currently'
+            ' can be CpuGpuBlockAllocator (the default) and '
+            'CpuOffloadingBlockAllocator (experimental) that '
+            'supports offloading the KV cache to CPU . '
+            'When using CpuOffloadingBlockAllocator, the '
+            'preemption mode must be recompute.')
```
Comment (comaniac) [inline vllm/engine/arg_utils.py:427]: Can we avoid to expose block allocator? Instead we should provide something like `--kv-cache-offloading`.

File: vllm/worker/worker.py [hunk 6]
```diff
@@ -322,10 +322,10 @@ def prepare_worker_input(
         # `blocks_to_swap_in` and `blocks_to_swap_out` are cpu tensors.
         # they contain parameters to launch cudamemcpyasync.
         blocks_to_swap_in = torch.tensor(execute_model_req.blocks_to_swap_in,
-                                         device="cpu",
+                                         device="cuda",
```
Comment (comaniac) [inline vllm/worker/worker.py:325]: This won't be compatible with other devices I think?

File: benchmarks/benchmark_long_document_qa.py
Comment (comaniac) [inline benchmarks/benchmark_long_document_qa.py:1]: Can we separate the benchmark scripts to another PR to reduce the size of this one?

File: csrc/cache_kernels.cu [hunk 7]
```diff
@@ -21,8 +22,64 @@
...
+template <typename T>
+T* get_kernel_ptr(torch::Tensor& tensor) {
+  // Get the kernel-accessible pointer of the given type T
+  // Returns NULL if the tensor is on CPU and non-pinned
+  torch::Device device = tensor.device();
+  if (device.is_cuda()) {
+    return static_cast<T*>(tensor.data_ptr());
+  } else if (device.is_cpu() && tensor.is_pinned()) {
+    T* ptr;
+    cudaHostGetDevicePointer((void**)&ptr, static_cast<T*>(tensor.data_ptr()),
```
Comment (ApostaC) [inline csrc/cache_kernels.cu:71]: Save this pointer at the creation of the CPU memory allocation -- making it CUDA graph compatible.

Comment (ApostaC) [reply csrc/cache_kernels.cu:71]: Potentially in a new PR

File: vllm/worker/worker.py (hunk 6 above)
Comment (ApostaC) [reply vllm/worker/worker.py:325]: As suggested by Kaichao, we will use a static pinned CPU memory buffer to host this array.

File: vllm/engine/arg_utils.py (hunk 5 above)
Comment (youkaichao) [reply vllm/engine/arg_utils.py:427]: After a second thought, for the sake of modularity and extensibility, I think it's fine to expose a block allocator argument, but we should keep the default name concise, and also keep the possibility to accept third-party allocators.

I recommend adding `--allocator`, which can take several default values as above, and also a class qualname like `mod.cls` .

Comment (comaniac) [reply vllm/engine/arg_utils.py:427]: How does this interface work with v1?

File: benchmarks/benchmark_long_document_qa.py
Comment (KuntaiDu) [reply benchmarks/benchmark_long_document_qa.py:1]: Yeah sure!

File: vllm/core/block/cpu_offloading_block_allocator.py [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+        cpu_allocator = self._allocators[Device.CPU]
+
+        new_uncached_blocks: Deque[Block] = deque()
+
+        while self._uncached_blocks:
+            block = self._uncached_blocks.pop()
+            block_id = block.block_id
+
+            # check if this block is freed
+            if block_id is None:
```
Comment (lixiaobai09) [inline vllm/core/block/cpu_offloading_block_allocator.py:329]: Because the block structure in _uncached_blocks is from a block memory pool, the block can be freed and allocated as a new block_id. However, while the block_id is allocated as a new block_id, this may cause some mistakes (or hurt performance). So, maybe we can remove the block in _uncached_blocks while block_manager calls the free function in cpu_offloading_block_allocator?

**Comment by ApostaC** [comment]
As suggested by @simon-mo and @youkaichao , I split this PR into 3 smaller PRs to make it easier to review. 
- CUDA kernel: #11531 
- CPU offloading block allocator: #11532 
- Benchmark script: #11533

File: vllm/core/block/cpu_offloading_block_allocator.py (hunk 8 above)
Comment (ApostaC) [reply vllm/core/block/cpu_offloading_block_allocator.py:329]: I see what you mean. This makes sense. Let me think what's the best way to fix it. Iterating over the `_uncache_blocks` during free() will be too expensive IMO.

Comment (ApostaC) [reply vllm/core/block/cpu_offloading_block_allocator.py:329]: @lixiaobai09 Hey, I've fixed this. Please see #11532

**Comment by hmellor** [comment]
After speaking to @ApostaC I'm closing this as it only optimises the V0 engine. There is work in progress for a similar PR in V1.
